src/ucc/api/ucc.h

```diff
@@ -199,33 +167,260 @@ typedef enum {
+#define   UCC_DT_UINT64   UCC_PREDEFINED_DT(8)
+#define   UCC_DT_UINT128  UCC_PREDEFINED_DT(9)
+#define   UCC_DT_FLOAT16  UCC_PREDEFINED_DT(10)
+#define   UCC_DT_FLOAT32  UCC_PREDEFINED_DT(11)
+#define   UCC_DT_FLOAT64  UCC_PREDEFINED_DT(12)
+#define   UCC_DT_PREDEFINED_LAST  13
+
+/**
+ * @ingroup UCC_DATATYPE
+ */
+enum ucc_generic_dt_ops_field {
+    UCC_GENERIC_DT_OPS_FIELD_FLAGS             = UCC_BIT(0),
+};
+
+/**
+ * @ingroup UCC_DATATYPE
+ * @brief Flags that can be specified for generic datatype
+ *
+ */
+
+typedef enum {
+    UCC_GENERIC_DT_OPS_FLAG_CONTIG             = UCC_BIT(0), /*!< If set, the created datatype
+                                                                  represents a contiguous memory
+                                                                  region with the size specified
+                                                                  in @ref contig_size field of
+                                                                  @ref ucc_generic_dt_ops */
+    UCC_GENERIC_DT_OPS_FLAG_REDUCE             = UCC_BIT(1), /*!< If set, the created datatype
+                                                                  has user-defined reduction
+                                                                  operation associated with it.
+                                                                  reduce.cb and reduce.ctx fields
+                                                                  of @ref ucc_generic_dt_ops must
+                                                                  be initialized. Collective operations
+                                                                  that involve reduction (allreduce,
+                                                                  reduce, reduce_scatter/v) can use
+                                                                  user-defined data-types only when
+                                                                  this flag is set. */
+} ucc_generic_dt_ops_flags_t;
+
+/**
+ * @ingroup UCC_DATATYPE
+ * @brief UCC generic data type descriptor
+ *
+ * This structure provides a generic datatype descriptor that
+ * is used to create user-defined datatypes.
+ */
+
+typedef struct ucc_generic_dt_ops {
+    uint64_t mask;
+    uint64_t flags;
+    size_t   contig_size; /*!< size of the datatype if @ref UCC_GENERIC_DT_OPS_FLAG_CONTIG is set */
+    /**
+     * @ingroup UCC_DATATYPE
+     * @brief Start a packing request.
+     *
+     * The pointer refers to application defined start-to-pack routine.
+     *
+     * @param [in]  context        User-defined context.
+     * @param [in]  buffer         Buffer to pack.
+     * @param [in]  count          Number of elements to pack into the buffer.
+     *
+     * @return  A custom state that is passed to the subsequent
+     *          @ref ucc_generic_dt_ops::pack "pack()" routine.
+     */
+    void* (*start_pack)(void *context, const void *buffer, size_t count);
```

Comment:
Did we decide to keep both start_pack and pack to have consistency with UCX or for some other reason?

Reply:
i believe consistency with UCX and easier integration at tl/ucp level were reasons during WG call.